File: experimentos/control/cd/Cajero/Estado.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Text.Json;

namespace Solucion;


/// <summary>
/// Modelo que representa el estado del sistema.
/// </summary>
public class Estado
{
    // Completar con las propiedades del estado.

    public IDictionary<int, Usuario> Usuarios { get; set; } 

    public IList<Movimiento> Movimientos { get; set; } 

    public decimal DineroDisponible { get; set; }

    public Dictionary<int, IList<RegistroCambioClave>> CambiosClave { get; set; }

    /// <summary>
    /// Guarda la instancia actual del modelo de Estado en el archivo `estado.json`.
    /// </summary>
    public void Guardar()
    {
        string ruta = "C:\\Users\\fs80235\\Desktop\\atm\\experimentos\\control\\Cepero-Diego\\Cajero\\estado.json";
        string contenido = JsonSerializer.Serialize(this);
        File.WriteAllText(ruta, contenido, Encoding.UTF8);
    }

    /// <summary>
    /// Retorna una instancia del modelo de estado con los valores guardados en `estado.json`.
    /// </summary>
    public static Estado Cargar()
    {
        string ruta = "C:\\Users\\fs80235\\Desktop\\atm\\experimentos\\control\\Cepero-Diego\\Cajero\\estado.json";
        string contenido = "";
        if (File.Exists(ruta))
        {
            contenido = File.ReadAllText(ruta, Encoding.UTF8);
        }

        if (contenido == "")
        {
            return Inicial();
        }
        else
        {
            return JsonSerializer.Deserialize<Estado>(contenido);
        }
    }


    /// <summary>
    /// Retorna una instancia del modelo de estado con sus valores iniciales.
    /// </summary>
    private static Estado Inicial()
    {
        var administrador = new Usuario()
        {
            NombreApellido = "Administrador",
            NumeroDocumento = 12345678, 
            EsAdministrador = true,
            HashClave = "adm1n1strad0r"
        };

        return new Estado()
        {
            Usuarios = new Dictionary<int, Usuario>
            {
                { administrador.NumeroDocumento, administrador },
            },
            Movimientos = new List<Movimiento>(),
            DineroDisponible = 0,
            CambiosClave = new()
        };
    }
}

/* --- Ejemplo de uso ---
Estado e = Estado.Cargar();
...
e.Guardar();
*/

File: experimentos/control/nm/estado.js
import { readFileSync, writeFileSync, existsSync } from "fs";
import { ATM } from "./atm.js";
import { Manager } from "./manager.js";

/** Modelo que representa el estado del sistema.
 * @class
 */
export class Estado {
  /**
   * @type {Manager} manager
   */
  manager

  /** Guarda la instancia actual del modelo de Estado en el archivo `estado.json`. */
  guardar() {
    writeFileSync("estado.json", JSON.stringify(this.manager), { encoding: "utf8" });
  }

  /** Retorna una instancia del modelo de estado con los valores guardados en `estado.json`. */
  static cargar() {
    let contenido = "";
    if (existsSync("estado.json")) {
      contenido = readFileSync("estado.json", { encoding: "utf8" }).toString();
    }

    if (contenido === "") {
      return Estado.#inicial();
    } else {
      const json = JSON.parse(contenido)
      const estado = new Estado()
      estado.manager = Manager.fromObject(json)
      return estado;
    }
  }
  /** Retorna una instancia del modelo de estado con sus valores iniciales. */
  static #inicial() {
    let estado = new Estado();
    estado.manager = new Manager(new ATM(), {}, {})
    return estado;
  }
}

/* --- Ejemplo de uso ---
import { Estado } from './estado.js';
let e = new Estado();
...
e.guardar();
*/
